core/packages/contracts/src/Gateway.sol

```diff
@@ -2,34 +2,445 @@
+
+        // Initialize an agent & channel for BridgeHub
+        Agent bridgeHubAgent = new Agent(params.bridgeHubAgentID);
+        $.agents[params.bridgeHubAgentID] = payable(bridgeHubAgent);
+        $.channels[params.bridgeHubParaID] = Channel({
+            mode: OperatingMode.Normal,
+            agent: address(bridgeHubAgent),
+            inboundNonce: 0,
+            outboundNonce: 0,
+            fee: params.fee,
+            reward: params.reward
+        });
+
+        // Initialize an agent & channel for AssetHub
+        Agent assetHubAgent = new Agent(params.assetHubAgentID);
+        $.agents[params.assetHubAgentID] = payable(assetHubAgent);
+        $.channels[params.assetHubParaID] = Channel({
+            mode: OperatingMode.Normal,
+            agent: address(assetHubAgent),
+            inboundNonce: 0,
+            outboundNonce: 0,
+            fee: params.fee,
+            reward: params.reward
+        });
+
+        Features.initialize(params.features, address(assetHubAgent));
+        Verification.initialize(params.verification);
+    }
+
+    function submitInbound(
+        InboundMessage calldata message,
+        bytes32[] calldata leafProof,
+        Verification.Proof calldata headerProof
+    ) external onlyProxy {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
+        Channel storage channel = ensureChannel(message.origin);
+
+        bytes32 leafHash = keccak256(abi.encode(message));
+        bytes32 commitment = MerkleProof.processProof(leafProof, leafHash);
+
+        if (!Verification.verifyCommitment(commitment, headerProof)) {
+            revert InvalidProof();
+        }
+
+        if (message.nonce != channel.inboundNonce + 1) {
+            revert InvalidNonce();
+        }
 
-    constructor(Registry registry) RegistryLookup(registry) {
-        _setRoleAdmin(SENDER_ROLE, ADMIN_ROLE);
+        // Increment nonce for origin.
+        channel.inboundNonce++;
+
+        // reward the relayer from the agent contract
+        // Should revert if there are not enough funds. In which case, the origin
+        // should top up the funds and have a relayer resend the message.
+        if (channel.reward > 0) {
+            Agent(payable(channel.agent)).withdrawTo(payable(msg.sender), channel.reward);
+        }
+
+        // Ensure relayers pass enough gas for message to execute.
+        // Otherwise malicious relayers can break the bridge by allowing handlers to run out gas.
+        // Resubmission of the message by honest relayers will fail as the tracked nonce
+        // has already been updated.
+        if (gasleft() < $.gasToForward + GAS_BUFFER) {
+            revert NotEnoughGas();
+        }
+
+        bool success = true;
+
+        if (message.command == COMMAND_AGENT_EXECUTE) {
+            try Gateway(this).handleAgentExecute{gas: $.gasToForward}(message.params) {}
+            catch {
+                success = false;
+            }
```

Comment:
Just curious why not wrap `handle*` as internal call here?

Reply:
The reason we use an external call is because its not possible to catch errors in internal calls using `try catch`.

If an internal call reverted, it would revert the entire transaction. And that could be problematic, as the channel would get blocked.

Reply:
Gotcha! Then will it make more sense if we make these `handle*` functions return some particular error code instead of the revert?  The error code could be also included in the event emitted.  In this way we can make these functions internal and also make client/relayer has more knowlege about the internal detail in ethereum side.